README.md
Quick Export - Export a vector layer to CSV, HTML, PDF or printer in 1 click
=============================================================================

This plugin has been developped by 3liz ( http://3liz.com )  .

    begin       : 2014-03-20
    copyright   : (C) 201 by 3liz
    authors     : René-Luc D'Hont and Michaël Douchin
    email       : info@3liz.com
    website     : http://www.3liz.com

Features
----------

Export tools
_____________

**Quick Export** adds a new toolbar in QGIS, with 4 actions to export any vector layer :

* **to CSV**: the exported file does not contains the geometries yet (may be a option in future versions). The delimiter can be set in the option dialog (comma, tab, semicolon or pipe).
* **to HTML**: the file can be opened in any browser, and displays the attribute table styled via CSS
* **to PDF**: the PDF contains the same table. It may be cut if too many fields are present in the attribute table. (a future version may correct this by dividing the table into pages when needed)
* **to default printer** : the PDF is generated in a temporary file, and the printer configuration dialog is opened. The user can select the printer and change the options.


Options dialog
________________

Some settings can be changed in the Options dialog, which can be opened via the fifth grey button [O] in the toolbar :

* General settings

  - Export hidden attributes : if checked, all the fields will be exported, even if marked as Hidden via the edit type option (fields tab of the vector layer properties dialog)

* CSV export

  - The delimiter can be chosen between tab, comma, pipe and semicolon


The Original Code is 3liz code.

Compatibility with QGIS 1.8
----------------------------

The plugin can be installed and used with QGIS 1.8 and above.

In QGIS 1.8, some features are a bit different :

* **PDF export** : the attribute table is not divided into one table per page, because of a bug in the related version of the QT lib. The lines of the tables may be cut in half between pages.
* **Communication with the user** : the message bar is only available since QGIS 2.0. In QGIS 1.8, messages boxes (smal dialogs with buttons) are shown instead to warn the user or let him open the file after a successful export.
* **Performance** : the export is faster with QGIS 2.0, thanks to changes in the api.

Authors
-------

The Initial Developer of the Original Code is Michaël Douchin <mdouchin@3liz.com>. Portions created by the Initial Developer are Copyright (C) 2014 the Initial Developer. All Rights Reserved.

Funders
--------------

The plugin has been initially funded by the **Conseil Général des Pyrénées-Atlantiques** (France) : http://www.cg64.fr


changelog.txt
CHANGELOG

  Version 1.2
  * Debug export for DATE and DATETIME format - bug #3
  * Debug wrong settings passed depending on language
  * Add a options dialog to let the user choose default behaviour : csv delimiter & hidden fields
  * Export only fields wich are not Hidden (fields tab of vector layer properties)
  * Export CSV via csv module instead of QgsVectorFileWriter.writeAsVectorFormat

  Version 1.1
  * Debug QGIS 1.8 export when no feature is selected
  * Print - open the printer configuration dialog
  * Debug date formatting in Windows - bug #2
  * Use CSS internally to avoid webkit bugs
  * Handle file deletion exception on Windows - bug #1

  Version 1.0
  * Initial version : https://github.com/3liz/QgisQuickExportPlugin


__init__.py
# -*- coding: utf-8 -*-
"""
/***************************************************************************
 QuickExport
                                 A QGIS plugin
 This plugin adds a toolbar with buttons to export the selected layer data to CSV, HTML and PDF files
                             -------------------
        begin                : 2014-03-17
        copyright            : (C) 2014 by 3liz
        email                : info@3liz.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
 This script initializes the plugin, making it known to QGIS.
"""

def classFactory(iface):
    # load QuickExport class from file QuickExport
    from quickexport import QuickExport
    return QuickExport(iface)


resources.py
# -*- coding: utf-8 -*-

# Resource object code
#
# Created: lun. mars 17 15:30:11 2014
#      by: The Resource Compiler for PyQt (Qt v4.8.4)
#
# WARNING! All changes made in this file will be lost!

from PyQt4 import QtCore

qt_resource_data = "\
\x00\x00\x04\x0a\
\x89\
\x50\x4e\x47\x0d\x0a\x1a\x0a\x00\x00\x00\x0d\x49\x48\x44\x52\x00\
\x00\x00\x17\x00\x00\x00\x18\x08\x06\x00\x00\x00\x11\x7c\x66\x75\
\x00\x00\x00\x01\x73\x52\x47\x42\x00\xae\xce\x1c\xe9\x00\x00\x00\
\x06\x62\x4b\x47\x44\x00\xff\x00\xff\x00\xff\xa0\xbd\xa7\x93\x00\
\x00\x00\x09\x70\x48\x59\x73\x00\x00\x0b\x13\x00\x00\x0b\x13\x01\
\x00\x9a\x9c\x18\x00\x00\x00\x07\x74\x49\x4d\x45\x07\xd9\x02\x15\
\x16\x11\x2c\x9d\x48\x83\xbb\x00\x00\x03\x8a\x49\x44\x41\x54\x48\
\xc7\xad\x95\x4b\x68\x5c\x55\x18\xc7\x7f\xe7\xdc\x7b\x67\xe6\xce\
\x4c\x66\x26\x49\xd3\x24\x26\xa6\xc6\xf8\x40\x21\xa5\x04\xb3\x28\
\xda\x98\x20\xa5\x0b\xad\x55\xa8\x2b\xc5\x50\x1f\xa0\x6e\x34\x2b\
\x45\x30\x14\x02\xba\x52\x69\x15\x17\x66\x63\x45\x97\x95\xa0\xad\
\x0b\xfb\xc0\x06\x25\xb6\x71\x61\x12\x41\x50\xdb\x2a\x21\xd1\xe2\
\x24\xf3\x9e\xc9\xcc\xbd\xe7\x1c\x17\x35\x43\x1e\x33\x21\xb6\xfd\
\x56\x87\xf3\x9d\xfb\xfb\x1e\xf7\xff\x9d\x23\x8c\x31\x43\x95\xf4\
\x85\x1e\x3f\x3b\x35\xac\xfd\xcc\x43\xdc\xa4\x49\x3b\xfe\x9d\x1d\
\xdb\x7b\x22\x90\x78\xf8\xb2\x28\xa7\xbe\x7d\xc1\x4b\x9d\x79\xdf\
\x18\x15\xe5\x16\x99\x10\x56\xde\x69\xdc\x3f\x22\xfd\xec\xd4\xf0\
\xad\x04\x03\x18\xa3\xa2\x7e\x76\x6a\x58\xde\x68\x2b\xb4\x36\xf8\
\xbe\xc6\x18\x53\xdb\xef\xe7\xfa\xec\xed\x67\x63\x10\x42\x00\xf0\
\xfb\xd5\x65\x2a\x15\x45\xc7\x6d\x0d\x00\xc4\xa2\xc1\xaa\x6f\x0d\
\x3e\x6c\xab\xc2\x1c\x56\xa4\x77\x4b\xb0\xf2\x35\x15\x5f\x21\x85\
\xe0\xc8\x6b\x5f\x92\x2d\x37\x33\x39\xf9\x03\x27\x8e\x1f\xa2\xf7\
\xbe\x9d\x04\x1c\x0b\x37\xe4\xac\xff\xa6\x30\x87\xbd\xba\x00\x6a\
\x06\x79\xe5\xf5\xaf\x89\xd9\x92\xc5\xcc\x0a\xd9\x7c\x19\xcf\xe9\
\xe2\xe4\xa9\x2f\x78\x7c\xff\x01\x72\x85\x0a\x2b\x65\x1f\xa5\x4c\
\xb5\xb2\x55\x16\x80\xbd\x31\xda\xda\x20\x1f\x7d\x3e\xcd\xc2\xfd\
\x59\xa6\x93\x39\x92\xd1\x22\xea\x9b\x16\xce\x9d\x3f\xce\xe0\x83\
\x03\x24\x82\x59\x3a\xdb\x7b\x88\xc7\x82\x68\x63\x58\xc9\xcc\x62\
\x8c\x21\x18\xb0\x6a\xc3\x37\x06\x49\x16\xff\x24\x6b\xa5\x49\xbb\
\x25\xbc\xa2\xa6\x21\xbb\x40\x7f\xdf\x00\x83\xbd\x01\x8e\x3c\xd5\
\x45\xd7\x8e\x6b\x9c\x9c\x98\x25\x1a\xb6\xe8\xbe\x3d\xc2\xdd\x77\
\x44\x48\xc4\x1c\x22\xe1\xeb\x58\x59\xaf\xcf\xd3\x33\x29\x2e\x34\
\x2d\x91\x93\x3e\xbe\x34\x78\x01\xc5\xe2\x61\xc5\xae\x72\x8e\x70\
\xc8\xc2\x0d\x5a\xbc\xf5\xee\x2f\x9c\xfa\x3e\x86\x69\x7a\x8e\xcf\
\x26\xe6\xf9\x63\xa1\x44\xa1\xa4\xd0\xda\x6c\x0d\x2f\x15\x7c\xb4\
\x67\x28\x59\x0a\xcf\xd6\x54\xe2\x06\x13\x87\x2b\x6f\x68\xa6\x27\
\xaf\x31\x32\x36\xc7\xb2\x7f\x17\xef\x7d\x7c\x8c\x33\x67\xcf\x12\
\x70\x24\x4a\x69\xd6\x6a\x46\xd6\xd3\x70\x72\xa9\x82\x67\x34\x45\
\xad\x28\xdb\x1a\x15\x34\x98\xff\x46\xed\xef\x37\x0d\x99\xbf\x4a\
\x3c\x30\x38\xc0\xc8\x4b\xaf\x92\x5a\x9c\xe2\xe0\x23\x6d\x74\xb4\
\xba\x84\x5d\x0b\x29\x45\x7d\xb8\x94\x82\x96\xb6\x10\xf3\xc5\x12\
\x2a\xef\x53\x11\x1a\x63\xad\x3f\x93\x19\x85\xf1\xb1\x77\x58\x5a\
\xf8\x99\x97\x9f\xe9\xa6\x75\x47\x90\xc6\xb8\x43\xd8\xb5\xb6\xce\
\xfc\xfa\xfd\x00\xfb\x3e\xf4\xc8\x05\x35\xba\x5e\xeb\x46\x21\xf9\
\xcf\x0a\xa9\x8c\x87\xe3\x48\xdc\x90\xb5\x6e\x98\x6a\xaa\x65\xf2\
\x52\x92\x43\x2f\x5e\xc2\x8c\x02\x1a\x10\xf5\x07\xac\xc3\x75\x70\
\x83\x92\x80\xb3\xf9\xd0\x26\xf8\x8f\xb3\x29\xc6\x3e\xb8\x8c\x19\
\x35\x75\x6b\x7b\x7e\x3c\xca\x45\x0c\x7e\x49\x31\xf4\x58\x3b\xf7\
\xf6\x34\x90\x88\x39\x04\x1c\x59\x1f\xfe\xdb\xd5\x3c\x5f\x9d\x4b\
\x32\xfd\x44\xb2\xba\xd7\xfa\xb6\x60\xcf\xde\x16\xdc\x90\x45\x4c\
\x4a\x2a\x9e\x62\xfe\x4e\xc5\xc8\xc1\x4e\xda\x76\x86\xe8\xe9\x0a\
\xe3\xd8\x92\x58\xd4\xc6\xb2\x44\x6d\x78\x2a\x53\xe1\xca\x7c\x99\
\x63\x5d\xbf\x56\x9d\xbd\x9f\x44\x18\x7a\xba\x95\x27\x0f\xb4\xd3\
\xdc\x18\xc0\xf3\x0d\x52\x40\xd8\xb5\xb0\xa4\x20\x14\xb2\x70\x6c\
\x81\x63\xcb\xaa\x42\xd6\xfd\xb7\xf4\xec\xa3\x06\xa0\x50\x52\xd8\
\x4e\x1b\x7e\x4a\xd3\x31\xf9\x29\xcf\xfe\xd4\x49\x7f\x5f\x13\xfb\
\xfa\x9b\x71\x43\x92\x58\xd4\x21\x18\x90\xac\xde\xb0\x42\x50\x13\
\x58\x33\xf3\x88\x6b\xa1\xfd\x65\x96\xf2\x79\xc6\x43\x7b\xd8\x75\
\x38\xcc\x3d\xdd\xd1\xaa\xcf\x71\xe4\xff\x7f\x91\x56\x33\xaf\xea\
\x37\xe7\xa1\x94\x21\x16\xb5\xd1\x06\x2c\x29\x36\xf5\x72\x9b\x96\
\x95\xc0\xc4\xda\x9d\x78\x83\x43\x53\x22\x80\x65\x09\x1c\xfb\x86\
\xc1\x00\xe7\x25\x70\x14\x48\x6f\x1e\x22\x51\xe3\x75\xd9\xb6\xa5\
\x81\xa3\x32\xb1\xfb\xf4\x0c\x30\xb8\xb1\x82\x9b\xb0\x09\x60\x30\
\xb1\xfb\xf4\xcc\xbf\xa0\xe9\x6e\xae\x5a\xdf\x4b\x81\x00\x00\x00\
\x00\x49\x45\x4e\x44\xae\x42\x60\x82\
"

qt_resource_name = "\
\x00\x07\
\x07\x3b\xe0\xb3\
\x00\x70\
\x00\x6c\x00\x75\x00\x67\x00\x69\x00\x6e\x00\x73\
\x00\x0b\
\x01\x3e\x09\xd4\
\x00\x71\
\x00\x75\x00\x69\x00\x63\x00\x6b\x00\x65\x00\x78\x00\x70\x00\x6f\x00\x72\x00\x74\
\x00\x08\
\x0a\x61\x5a\xa7\
\x00\x69\
\x00\x63\x00\x6f\x00\x6e\x00\x2e\x00\x70\x00\x6e\x00\x67\
"

qt_resource_struct = "\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x14\x00\x02\x00\x00\x00\x01\x00\x00\x00\x03\
\x00\x00\x00\x30\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\
"

def qInitResources():
    QtCore.qRegisterResourceData(0x01, qt_resource_struct, qt_resource_name, qt_resource_data)

def qCleanupResources():
    QtCore.qUnregisterResourceData(0x01, qt_resource_struct, qt_resource_name, qt_resource_data)

qInitResources()


quickexportdialog.py
# -*- coding: utf-8 -*-
"""
/***************************************************************************
 QuickExportDialog
                                 A QGIS plugin
 This plugin adds a toolbar with buttons to export the selected layer data to CSV, HTML and PDF files
                             -------------------
        begin                : 2014-03-17
        copyright            : (C) 2014 by 3liz
        email                : info@3liz.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
"""

from PyQt4 import QtCore, QtGui
from ui_quickexport import Ui_QuickExport
# create the dialog for zoom to point


class QuickExportDialog(QtGui.QDialog, Ui_QuickExport):
    def __init__(self):
        QtGui.QDialog.__init__(self)
        # Set up the user interface from Designer.
        # After setupUI you can access any designer object by doing
        # self.<objectname>, and you can use autoconnect slots - see
        # http://qt-project.org/doc/qt-4.8/designer-using-a-ui-file.html
        # #widgets-and-dialogs-with-auto-connect
        self.setupUi(self)


quickexport_option_form.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'quickexport_option_form.ui'
#
# Created: Fri Apr  4 15:06:30 2014
#      by: PyQt4 UI code generator 4.10.3
#
# WARNING! All changes made in this file will be lost!

from PyQt4 import QtCore, QtGui

try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    def _fromUtf8(s):
        return s

try:
    _encoding = QtGui.QApplication.UnicodeUTF8
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig)

class Ui_quickexport_option_form(object):
    def setupUi(self, quickexport_option_form):
        quickexport_option_form.setObjectName(_fromUtf8("quickexport_option_form"))
        quickexport_option_form.resize(570, 242)
        self.verticalLayout = QtGui.QVBoxLayout(quickexport_option_form)
        self.verticalLayout.setObjectName(_fromUtf8("verticalLayout"))
        self.scrollArea = QtGui.QScrollArea(quickexport_option_form)
        self.scrollArea.setWidgetResizable(True)
        self.scrollArea.setObjectName(_fromUtf8("scrollArea"))
        self.scrollAreaWidgetContents = QtGui.QWidget()
        self.scrollAreaWidgetContents.setGeometry(QtCore.QRect(0, 0, 550, 222))
        self.scrollAreaWidgetContents.setObjectName(_fromUtf8("scrollAreaWidgetContents"))
        self.verticalLayout_2 = QtGui.QVBoxLayout(self.scrollAreaWidgetContents)
        self.verticalLayout_2.setObjectName(_fromUtf8("verticalLayout_2"))
        self.groupBox_3 = QtGui.QGroupBox(self.scrollAreaWidgetContents)
        self.groupBox_3.setObjectName(_fromUtf8("groupBox_3"))
        self.verticalLayout_5 = QtGui.QVBoxLayout(self.groupBox_3)
        self.verticalLayout_5.setObjectName(_fromUtf8("verticalLayout_5"))
        self.gridLayout_2 = QtGui.QGridLayout()
        self.gridLayout_2.setObjectName(_fromUtf8("gridLayout_2"))
        self.label_7 = QtGui.QLabel(self.groupBox_3)
        self.label_7.setObjectName(_fromUtf8("label_7"))
        self.gridLayout_2.addWidget(self.label_7, 0, 0, 1, 1)
        self.cbExportHiddenAttributes = QtGui.QCheckBox(self.groupBox_3)
        self.cbExportHiddenAttributes.setText(_fromUtf8(""))
        self.cbExportHiddenAttributes.setObjectName(_fromUtf8("cbExportHiddenAttributes"))
        self.gridLayout_2.addWidget(self.cbExportHiddenAttributes, 0, 1, 1, 1)
        self.verticalLayout_5.addLayout(self.gridLayout_2)
        self.verticalLayout_2.addWidget(self.groupBox_3)
        self.groupBox = QtGui.QGroupBox(self.scrollAreaWidgetContents)
        self.groupBox.setObjectName(_fromUtf8("groupBox"))
        self.verticalLayout_3 = QtGui.QVBoxLayout(self.groupBox)
        self.verticalLayout_3.setObjectName(_fromUtf8("verticalLayout_3"))
        self.gbDelimiter = QtGui.QGroupBox(self.groupBox)
        self.gbDelimiter.setTitle(_fromUtf8(""))
        self.gbDelimiter.setAlignment(QtCore.Qt.AlignLeading|QtCore.Qt.AlignLeft|QtCore.Qt.AlignTop)
        self.gbDelimiter.setFlat(False)
        self.gbDelimiter.setCheckable(False)
        self.gbDelimiter.setObjectName(_fromUtf8("gbDelimiter"))
        self.horizontalLayout = QtGui.QHBoxLayout(self.gbDelimiter)
        self.horizontalLayout.setSpacing(6)
        self.horizontalLayout.setContentsMargins(-1, 0, -1, -1)
        self.horizontalLayout.setObjectName(_fromUtf8("horizontalLayout"))
        self.label = QtGui.QLabel(self.gbDelimiter)
        font = QtGui.QFont()
        font.setBold(True)
        font.setWeight(75)
        self.label.setFont(font)
        self.label.setObjectName(_fromUtf8("label"))
        self.horizontalLayout.addWidget(self.label)
        self.rbDelimiterTab = QtGui.QRadioButton(self.gbDelimiter)
        self.rbDelimiterTab.setChecked(True)
        self.rbDelimiterTab.setObjectName(_fromUtf8("rbDelimiterTab"))
        self.horizontalLayout.addWidget(self.rbDelimiterTab)
        self.rbDelimiterComma = QtGui.QRadioButton(self.gbDelimiter)
        self.rbDelimiterComma.setObjectName(_fromUtf8("rbDelimiterComma"))
        self.horizontalLayout.addWidget(self.rbDelimiterComma)
        self.rbDelimiterPipe = QtGui.QRadioButton(self.gbDelimiter)
        self.rbDelimiterPipe.setObjectName(_fromUtf8("rbDelimiterPipe"))
        self.horizontalLayout.addWidget(self.rbDelimiterPipe)
        self.rbDelimiterSemicolon = QtGui.QRadioButton(self.gbDelimiter)
        self.rbDelimiterSemicolon.setObjectName(_fromUtf8("rbDelimiterSemicolon"))
        self.horizontalLayout.addWidget(self.rbDelimiterSemicolon)
        self.verticalLayout_3.addWidget(self.gbDelimiter)
        self.verticalLayout_2.addWidget(self.groupBox)
        spacerItem = QtGui.QSpacerItem(20, 40, QtGui.QSizePolicy.Minimum, QtGui.QSizePolicy.Expanding)
        self.verticalLayout_2.addItem(spacerItem)
        self.buttonBox = QtGui.QDialogButtonBox(self.scrollAreaWidgetContents)
        self.buttonBox.setStandardButtons(QtGui.QDialogButtonBox.Cancel|QtGui.QDialogButtonBox.Ok)
        self.buttonBox.setObjectName(_fromUtf8("buttonBox"))
        self.verticalLayout_2.addWidget(self.buttonBox)
        self.scrollArea.setWidget(self.scrollAreaWidgetContents)
        self.verticalLayout.addWidget(self.scrollArea)

        self.retranslateUi(quickexport_option_form)
        QtCore.QMetaObject.connectSlotsByName(quickexport_option_form)
        quickexport_option_form.setTabOrder(self.scrollArea, self.buttonBox)

    def retranslateUi(self, quickexport_option_form):
        quickexport_option_form.setWindowTitle(_translate("quickexport_option_form", "QuickExport - Settings", None))
        self.groupBox_3.setTitle(_translate("quickexport_option_form", "General settings", None))
        self.label_7.setText(_translate("quickexport_option_form", "Export hidden attributes", None))
        self.groupBox.setTitle(_translate("quickexport_option_form", "CSV export", None))
        self.label.setText(_translate("quickexport_option_form", "Delimiter", None))
        self.rbDelimiterTab.setText(_translate("quickexport_option_form", "tab", None))
        self.rbDelimiterComma.setText(_translate("quickexport_option_form", "comma", None))
        self.rbDelimiterPipe.setText(_translate("quickexport_option_form", "pipe", None))
        self.rbDelimiterSemicolon.setText(_translate("quickexport_option_form", "semicolon", None))

import resource_rc
